demos/http/http_demo_s3_upload/http_demo_s3_upload.c

```diff
@@ -0,0 +1,626 @@
+
+    /* ALPN is required when communicating to AWS IoT Core over port 443 through HTTP. */
+    if( HTTPS_PORT == 443 )
+    {
+        opensslCredentials.pAlpnProtos = IOT_CORE_ALPN_PROTOCOL_NAME;
+        opensslCredentials.alpnProtosLen = strlen( IOT_CORE_ALPN_PROTOCOL_NAME );
+    }
+
+    /* Retrieve the address location and length from the S3_PRESIGNED_PUT_URL. */
+    httpStatus = getUrlAddress( S3_PRESIGNED_PUT_URL,
+                                strlen( S3_PRESIGNED_PUT_URL ),
+                                &pAddress,
+                                &serverHostLength );
+
+    if( httpStatus != HTTP_SUCCESS )
+    {
+        LogError( ( "An error occurred in getUrlAddress() on URL %s\r\n. Error code %d",
+                    S3_PRESIGNED_PUT_URL,
+                    httpStatus ) );
+        returnStatus = EXIT_FAILURE;
+    }
+
+    /* serverHost should consist only of the host address located in S3_PRESIGNED_PUT_URL. */
+    memcpy( &serverHost, pAddress, serverHostLength );
+
+    /* Initialize server information. */
+    serverInfo.pHostName = &serverHost;
+    serverInfo.hostNameLength = serverHostLength;
+    serverInfo.port = HTTPS_PORT;
+
+    /* Establish a TLS session with the HTTP server. This example connects
+     * to the HTTP server as specified in SERVER_HOST and HTTPS_PORT
+     * in demo_config.h. */
+    LogInfo( ( "Establishing a TLS session with %s:%d.",
+               &serverHost,
+               HTTPS_PORT ) );
+
+    opensslStatus = Openssl_Connect( pNetworkContext,
+                                     &serverInfo,
+                                     &opensslCredentials,
+                                     TRANSPORT_SEND_RECV_TIMEOUT_MS,
+                                     TRANSPORT_SEND_RECV_TIMEOUT_MS );
+
+    if( opensslStatus == OPENSSL_SUCCESS )
+    {
+        returnStatus = EXIT_SUCCESS;
+    }
+    else
+    {
+        returnStatus = EXIT_FAILURE;
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static HTTPStatus_t getS3ObjectFileSize( size_t * pFileSize,
+                                         const TransportInterface_t * pTransportInterface,
+                                         const char * pHost,
+                                         const char * pMethod,
+                                         const char * pPath )
+{
+    HTTPStatus_t httpStatus = HTTP_SUCCESS;
+
+    /* The location of the file size in contentRangeValStr. */
+    char * pFileSizeStr = NULL;
+
+    /* String to store the Content-Range header value. */
+    char * contentRangeValStr = NULL;
+    size_t contentRangeValStrLength = 0;
+
+    /* Initialize the request object. */
+    requestInfo.pHost = pHost;
+    requestInfo.hostLen = strlen( pHost );
+    requestInfo.method = pMethod;
+    requestInfo.methodLen = strlen( pMethod );
+    requestInfo.pPath = pPath;
+    requestInfo.pathLen = strlen( pPath );
```

Comment:
I would recommend a structure for a string and its length since we cannot use `strlen`:
```
/**
 * @brief A pair containing a path string of the URI and its length.
 */
typedef struct httpPathStrings
{
    const char * httpPath;
    size_t httpPathLength;
} httpPathStrings_t;

/**
 * @brief A pair containing an HTTP method string and its length.
 */
typedef struct httpMethodStrings
{
    const char * httpMethod;
    size_t httpMethodLength;
} httpMethodStrings_t;
```

Reply:
Or in other words checkout what the other HTTP demos do

Reply:
If each private function has a specific use case with no string parameters (in other words it knows exactly which string to reference, similar to mutual auth, and dissimilar to other demos where `sendHttpRequest` is used to send an array of different requests), and each string is only referenced once, are defining structs necessary? I had attempted to implement them but hadn't found much value in doing so, but if you think it would still be beneficial to show their use, I can definitely add them in.

Reply:
Oh yea sorry, I see that this path is the rest of the presigned URL.